Long and short packet header

janaiyengar opened the pull request:
This PR proposes an alternative header format for QUIC.  It is based on a previous proposal and incorporates feedback received on the list. This PR is unfortunately quite invasive.

Closes:
#35: Starting packet number
#56: Extending flags
#119: Server-proposed connection ID
#133: Connection ID in version negotiation
#135: DoS using Version Negotiation Packets
#147: Reflection Attack Resistance
#185: Reliable identification of the initial packet for a connection
#193: Flags section is kind of confusing
#244: Need a NONCE in version negotiation packets
#293: Does the connection ID need to be in a consistent location
#295: Connection ID on a version negotiation packet

(also addresses some of:
#148: QUIC packet header complexity
#170: Connection ID collisions)

alagoutte commented:
LGTM

ekr commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+* Octets 1-8: Connection ID (randomly chosen)
+* Octets 9-12: Packet number (low 4 octets, starts at a random 31-bit value)
+* Octets 13-16: Version
+* Octets 17+: Payload
```
I feel like cleartext would be better to go first. Lets do the happy path.

janaiyengar replied:
I'm doing it in the order that the packet types are defined. I've moved the types and the sections to have happy path first.

ekr commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+IDs are used after a successful handshake, cleanly distinguishing packets that
+use them from packets using client-selected Connection IDs. All packets
+including and following a Server Cleartext End-of-Handshake packet use a
+server-selected Connection ID, as described in {{connection-id}}.
```
I would consider not interleaving the description of the semantics and the packet format. The packet format tells you which bytes are bytes, so you just need the semantics here.

MikeBishop replied:
+1.  Define the fields, then describe how they're used by reference to those fields, not the underlying bytes.

martinthomson replied:
This is an artifact of me partially fixing the formatting.

janaiyengar replied:
Simplified all this significantly, and moved descriptions down to prose below the layouts.

ekr commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+as the Connection ID in all Cleartext packets and all 0-RTT Encrypted
+packets. The client's Connection ID is a suggestion to the server. The server
+echoes this value in all packets until the handshake is successful (see
+{{QUIC-TLS}}. In its final successful handshake completion packet, the server
```
"handshake completion packet" does not seem to be a defined term.

MikeBishop replied:
And are there multiple successful handshake completion packets?  Let's reference the specific packet type.

martinthomson replied:
I had the same question as @MikeBishop here.  I believe that the answer is that the server sends the "Final" long packet type on all of the packets after ServerHello.  It would only echo the client's connection ID if it needed to send a Version Negotiation packet or a HelloRetryReqest.

janaiyengar replied:
clarified and changed text. Yes there can be several Final packets, and the server-selected value is used for all of them.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+Short Packet Type:
+
+: The remaining 5 bits of octet 0 include one of 32 packet types.
+  {{short-packet-types}} lists the types that are defined for short packets.
```
We could unify the spaces, but there will be types that are only valid for long (public reset, version negotiation), and possibly types that are only valid for short.  If we did unify them, though, we could reclaim the key phase bit and just use the same 1-RTT (KP0) and 1-RTT (KP1) types the long header defines.

janaiyengar replied:
Yeah, there will be holes or overlap between the header types. I don't see the value in unifying since there's got to be a branch before reading the types about which bits to read from the packet anyways, and two small maps aren't really that expensive to maintain.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+|0|C|K|  Type   |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                                                               |
++                   Connection ID (optional)                    +
```
We have a notation convention for optional fields:  This would be [Connection ID (64)].

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+| 01   | 1 octet                |
+| 02   | 2 octets               |
+| 03   | 4 octets               |
+{: #short-packet-types title="Short Header Packet Types"}
```
What bit are they burning?  This appears to only be indicated by the type byte.

janaiyengar replied:
Not sure what this comment means

martinthomson replied:
Can't help you there.

MikeBishop replied:
Ekr had a comment (deleted?) that talked about both burning a bit _and_ using a frame type to communicate... something.  Packet number length, maybe?  My comment was that there no longer appears to be a dedicated bit, so I think his comment was invalid.  If he deleted it subsequently, apparently he thought so too.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+{: #short-packet-types title="Short Header Packet Types"}
+
+The header form, connection ID flag and connection ID of a short header packet
+are version independent.  The remaining fields are specific to the selected QUIC
```
Hyphenate:  version-independent.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+* Octets 1-8: Connection ID (echoed)
+* Octets 9-12: Packet Number (echoed)
+* Octets 13-16: Version (echoed)
+* Octets 17+: Payload (version list, containing 0 or more acceptable versions)
```
Yeah, it seems a little odd to allow the server to respond "I speak just enough QUIC to say 'I don't speak QUIC.'"

martinthomson replied:
I thinking that this is just @janaiyengar trolling us all.

janaiyengar replied:
Good point. Changed to 1 or more.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+{: #version-negotiation-format title="Version Negotiation Packet"}
+
+A server that generates a version negotiation packet SHOULD include one of the
+reserved version numbers (those following the pattern 0x?a?a?a?a).  This ensures
```
"helps ensure." Nothing can ensure well-written clients.

martinthomson replied:
Agreed, in spades, though this isn't new text for this PR.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+number.  This is because the server is required to include the same value in its
+transport parameters (see {{version-validation}}).  To avoid the selected
+version number changing during connection establishment, the reserved version
+can be generated as a function of values that will be available to the server
```
SHOULD?

martinthomson replied:
Not really.  The requirement here is that the bogus version is constant.  This is only suggesting one way that a server might be implemented to comply with that requirement.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+
+## Public Reset Packet {#public-reset-packet}
+
+A Public Reset packet is sent by only a server and is used to abruptly terminate
```
"only sent by servers"

martinthomson replied:
Hmm, I don't remember this being a restriction on its use.  Why only servers?

MikeBishop replied:
That was my initial reaction as well.  A client implementation on top of a UDP stack will never receive a packet on a closed connection, because there's no UDP socket open there.  The UDP stack will drop packets before they reach QUIC.  However, a server will have a listening socket open and receive all traffic that's sent to that port -- new connections, existing connections, and stale connections.

janaiyengar replied:
changing to "sent only by a server"

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+* Octets 13-16: Version
+* Octets 17+: Payload
+
+A Server Cleartext End-of-Handshake packet contains:
```
You called this Final above, not End-of-Handshake.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+* Octets 13-16: Version
+* Octets 17+: Encrypted Payload
+
+A flags octet of 0x86 indicates a 0-RTT packet. Key phases are used by the QUIC
```
You're pivoting away from 0-RTT without lead-in.  I'd suggest "...indicates a 0-RTT packet.  After the 1-RTT keys are established, key phases...."

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+packets. The client's Connection ID is a suggestion to the server. The server
+echoes this value in all packets until the handshake is successful (see
+{{QUIC-TLS}}. In its final successful handshake completion packet, the server
+MUST use a server-selected Connection ID. All subsequent packets from the server
```
Wait, if the Connection ID is a suggestion to the server, but the server MUST use a server-selected ID, what's the point of the suggestion?

Perhaps that packet type MUST contain the final Connection ID, which MAY be the one proposed by the client, or MAY be a new server-selected value?

martinthomson replied:
Yeah, I would just say any value.  The fact that the server might pick the value that the client did isn't particularly interesting.  If anything, we might point out that the ID doesn't have to be *new*, just unique.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -541,12 +721,54 @@ packet number.  Once any packet has been acknowledged, subsequent packets can
+Between different versions the following things are guaranteed to remain
+constant:
+
+* the location and size of the Flags field in both header forms,
```
You didn't originally call that field Flags in this version, though I see it has cropped up other places through the document.

martinthomson replied:
I would simply refer to this as the first octet, which sorts the both the "location" and "size" aspects.  Maybe:

> New versions of QUIC can be specified as long as the invariants in this section are maintained.
>
> The first octet always indicates the type of packet:
>  * The header form bit (0x80) always indicates a long or short header.
>  * In the short header, the next bit (0x40) indicates whether the Connection ID field is present.
>  * The remaining bits of this octet are interpreted as determined by the protocol version.
>
> Between all protocol versions, the following is guaranteed to be constant:
> * The location, size, and semantics of the Connection ID in both header forms (conditional on the field being present in the short form),
> * The location, size, and semantics of the Version field in the long form header, and
> * The location and size (but not semantics) of the Packet Number field in the long form header.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -541,12 +721,54 @@ packet number.  Once any packet has been acknowledged, subsequent packets can
+
+* the location and size of the Connection ID field in both header forms,
+
+* the location and value of the Version field in long headers, and
```
The value of the Version field can't change between versions?  That's really going to screw with version negotiation logic....  :-)  I suspect this should be location and meaning.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -541,12 +721,54 @@ packet number.  Once any packet has been acknowledged, subsequent packets can
+
+* the location and value of the Version field in long headers, and
+
+* the location and value of the Packet Number field in long headers.
```
Likewise.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -541,12 +721,54 @@ packet number.  Once any packet has been acknowledged, subsequent packets can
+* the location and value of the Packet Number field in long headers.
+
+Implementations MUST assume that an unsupported version uses an unknown packet
+format. All other values MUST be ignored when processing a packet that contains
```
Other fields, not just other values.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -541,12 +721,54 @@ packet number.  Once any packet has been acknowledged, subsequent packets can
+
+# Frames and Frame Types {#frames}
+
+The payload of cleartext packets and the plaintext after decryption of encrypted
```
After decryption of the payload -- you're not decrypting the entire packet.

martinthomson replied:
And let us talk about "(successful) removal of packet protection"

ekr commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+| 05   | Other Server Cleartext | {{cleartext-packet}}           |
+| 06   | 0-RTT Encrypted        | {{encrypted-packet}}           |
+| 07   | 1-RTT Encrypted (key phase 0) | {{encrypted-packet}}    |
+| 08   | 1-RTT Encrypted (key phase 1) | {{encrypted-packet}}    |
```
Oh, I see that I missed this in the short header. nm.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
 least significant bit is referred to as bit 0.  Hexadecimal notation is used for
 describing the value of fields.
 
+QUIC packet headers can be separated into long and short headers. Long headers
```
Done

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
 describing the value of fields.
 
+QUIC packet headers can be separated into long and short headers. Long headers
+are expected to be used for the initial phase of the connection and are used for
```
s/expected to be used/used/

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
 
-The fields in the Common Header are the following:
+Long headers are used for packets that are sent prior to the completion of
+version negotiation or are not encrypted with 1-RTT keys. Once both conditions
```
"prior to the completion of both version negotiation and the cryptographic handshake "

janaiyengar replied:
There's no way to send those packets otherwise, so this is definitional. I don't think we need to mandate behavior when no alternative is present.

martinthomson replied:
It's not definitional if you accept your example of out-of-band keying.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
 
+: The second bit (0x40) of octet 0 indicates whether the connection ID field is
+  present.  If set to 1, then the connection ID field is present; if set to 0,
+  the connection ID field is zero length.
```
same end result, no?

janaiyengar replied:
As mentioned on the list, and in agreement with Mike, there's value in having the packet be self-descriptive for connection ID.

janaiyengar replied:
changed to "omitted"

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+version negotiation, connection establishment, 0-RTT, and public reset
+packets. Short headers are minimal version-specific headers, which can be used
+after version negotiation and 1-RTT keys are established, and MUST NOT be used
+earlier.
```
Done

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
 
-   * 0x02 = PUBLIC_RESET.  Set to indicate that the packet is a Public Reset
-     packet.
+: The most significant bit (0x80) of the first octet (octet 0) is set to 1 for
```
Done

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
-     + 01 indicates that 2 bytes of the packet number are present
-     + 00 indicates that 1 byte of the packet number is present
+: Octets 1 through 8 include a connection ID.  {{connection-id}} describes the
+  use of this field in more details.
```
changed to "contain the connection ID"

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
 
-   * 0x80 is currently unused, and must be set to 0.
+: Octets 9 to 12 contain a packet number.  {{packet-numbers} describes the use
+  of packet numbers.
```
Done

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
 
+: Octets 13 to 16 contain the selected protocol version.  This field indicates
+  which version of QUIC is in use and determines how the some protocol fields
+  are interpreted.
```
done.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
-packets: Regular packets, Version Negotiation packets, and Public Reset packets.
-The flowchart below shows how a packet is classified into one of these three
-packet types:
+: Octets from 17 onwards are the payload of the packet.
```
I've added "(the rest of the QUIC packet)" for consistency since we don't really need to talk about UDP datagram anywhere else after QUIC packet is defined.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+| 02   | Public Reset           | {{public-reset-packet}}        |
+| 03   | Client Cleartext       | {{cleartext-packet}}           |
+| 04   | Final Server Cleartext | {{cleartext-packet}}           |
+| 05   | Other Server Cleartext | {{cleartext-packet}}           |
```
Changed consistently to Final and Non-Final

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+{: #long-packet-types title="Long Header Packet Types"}
 
-### Handling Packets from Different Versions
+The header form, long packet type, connection ID, packet number and version
```
I've added a sentence saying the packet types defined here are version-specific. I'm not sure about this though, so I've added a TODO as well.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+
+A pseudorandom function that takes client address information (IP and port) and
+the client selected version as input would ensure that there is sufficient
+variability in the values that a server uses.
```
moved this down to version negotiation section in a subsection called "Using Reserved Versions"

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+variability in the values that a server uses.
+
+A client MAY send a packet using a reserved version number.  This can be used to
+solicit a list of supported versions from a server.
```
moved

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+
+### Public Reset Proof
+
+Details to be added.
```
done

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -322,175 +322,356 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+communications. Public Reset is provided as an option of last resort for a
+server that does not have access to the state of a connection.  This is intended
+for use by a server that has lost state (for example, through a crash or
+outage). A server that wishes to indicate fatal errors with a connection MUST
```
done

martinthomson commented:
I think that we should just land this now:

![land it](https://images.duckduckgo.com/iu/?u=http%3A%2F%2Fwww.jokeoftheday.com%2Fpictures%2Fpymtrsi1ndxq7zj0ve3kfh9a86ugl4wob52c_Badlanding.jpg&f=1)

janaiyengar commented:
I've addressed all comments, PTAL. I hope to land this Thu/Fri, and I hope it's state is better than mt's plane :-)

ekr commented on draft-ietf-quic-transport.md:
```diff
@@ -336,175 +336,327 @@ all field sizes are in bits.  When discussing individual bits of fields, the
+location in all packet headers, making it straightforward for middleboxes, such
+as load balancers, to locate and use it.
 
+When a connection is initiated, the client MUST choose a random value and use it
```
What is the purpose of this being random?

janaiyengar replied:
The client then provides a suggestion that may be reasonably for servers to simply use. Only a server that cares about structure in the ID changes it, others use the clients suggested value as is.

janaiyengar commented:
Mike, you still have "Changes Requested".  I've addressed your comments though, so I'll go ahead and merge. Feel free to let me know if there's more you'd like me to do, happy to do it in a follow up PR.
